src/main/java/frc/robot/subsystems/Drivetrain.java

```diff
@@ -1,65 +1,525 @@
+      }
+    }
+  }
+  
+
+  public void setChassisSpeeds(ChassisSpeeds chassisSpeeds) {
+    m_swerveModuleStates = m_kinematics.toSwerveModuleStates(chassisSpeeds);
+    SwerveDriveKinematics.desaturateWheelSpeeds(m_swerveModuleStates, DriveConstants.kMaxSpeed);
+    setModuleStates(m_swerveModuleStates);
+  }
+
+  public PIDController getXController() {
+      return m_xController;
+  }
+  public PIDController getYController() {
+      return m_yController;
+  }
+  public PIDController getRotationController() {
+    return m_rotationController;
+  }
+  public void setAllOptimize(Boolean optimizeSate){
+    for (int i = 0; i < 4; i++) {
+      m_modules[i].setOptimize(optimizeSate);
+    }
+  }
+  public void setupDrivetrainShuffleboard() {
+    // inputs
+    m_heading = m_drivetrainTab.add("Set Heading (-pi to pi)", 0).getEntry();
+    
+    // add PID controlers
+    m_drivetrainTab.add("xController", getXController());
+    m_drivetrainTab.add("yController", getYController());
+    m_drivetrainTab.add("rotationController", getRotationController());
+
+    m_drivetrainTab.addNumber("getAngle", () -> getAngleHeading());
+    m_drivetrainTab.addNumber("heading PID output", () -> m_headingPIDOutput);
+
+    m_drivetrainTab.addNumber("Gyro X", () -> getAngularRate(0));
+    m_drivetrainTab.addNumber("Gyro Y", () -> getAngularRate(1));
+    m_drivetrainTab.addNumber("Gyro Z", () -> getAngularRate(2));
+
+    // m_drivetrainTab.add("odometry", m_odometry);
+
+    m_drivetrainTab.add(getXController());
+    m_drivetrainTab.add(getYController());
+    m_drivetrainTab.add(getRotationController());
+  }
+  public void setupModulesShuffleboard(){
+    setUpModuleChooser();
+    setUpFeedforwardSavers();
+    
+    // inputs
+    m_swerveModulesTab.add("Module Chooser", m_moduleChooser);
+    m_driveVelocity = m_swerveModulesTab.add("Set Drive Velocity", 0).getEntry();
+    m_steerVelocity = m_swerveModulesTab.add("Set Steer Velocity", 0).getEntry();
+    m_steerAngle = m_swerveModulesTab.add("Set Steer Angle", 0).getEntry();
+    m_drivetrainvolts = m_swerveModulesTab.add("Set Volts", 0).getEntry();
+    m_driveStaticFeedforward = m_swerveModulesTab.add("Drive kS FF", m_driveStaticFeedForwardSaver[m_moduleChooser.getSelected().getModuleType().getID()]).getEntry();
+    m_driveVelocityFeedforward = m_swerveModulesTab.add("Drive kV FF", m_driveVelFeedForwardSaver[m_moduleChooser.getSelected().getModuleType().getID()]).getEntry();
+    m_steerStaticFeedforward = m_swerveModulesTab.add("Steer kS FF", m_steerStaticFeedForwardSaver[m_moduleChooser.getSelected().getModuleType().getID()]).getEntry();
+    m_steerVelocityFeedforward = m_swerveModulesTab.add("Steer kV k FF", m_steerVelFeedForwardSaver[m_moduleChooser.getSelected().getModuleType().getID()]).getEntry();
+    
+
+    
+    
+    for (int i = 0; i < 4; i++){
+      m_modules[i].setupModulesShuffleboard();
+    }
+  }
+  private void setUpFeedforwardSavers(){
```

Comment:
The saver functionality introduces significant complexity to this module. Do we actually need it?  Why not simply update the constants based on the characterize output?

Reply:
the reason why we put it on shuffleboard is so we can copy paste it without using the log as it can be kind of finicky

Reply:
OK,  looks you strongly prefer to use Shuffleboard to output characterization results.  But then, do we have to have these savers to be inside  Drivetrain.java? I suggest we put them in the characterization code itself, or inside the callers of the charaterization command.

Reply:
Because the modules need the entry constantly we can't put the shuffleboard entry or savers in the command or the caller.